pkg/resourcevalidator/info.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2019-2022 The Liqo Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resourcevalidator
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	v1 "k8s.io/api/core/v1"
+)
+
+type peeringInfo struct {
+	clusterID        string
+	clusterName      string
+	SPList           map[string]ShadowPodDescription
+	PeeringQuota     v1.ResourceList
+	FreePeeringQuota v1.ResourceList
+	piMutex          sync.RWMutex
+	LastUpdateTime   string
+	// running          bool
+}
+
+/**
+ * PeeringInfo methods
+ */
+
+func (pi *peeringInfo) Lock() {
+	pi.piMutex.Lock()
+}
+
+func (pi *peeringInfo) Unlock() {
+	pi.piMutex.Unlock()
+}
+
+/* func (pi *peeringInfo) isRunning() bool {
+	return pi.running
+}
+
+func (pi *peeringInfo) terminate() {
+	pi.Lock()
+	pi.running = false
+	pi.Unlock()
+}
+
+func (pi *peeringInfo) start() {
+	pi.Lock()
+	pi.running = true
+	pi.Unlock()
+} */
+
+func getOrCreatePeeringInfo(cache *peeringCache, clusterID, clusterName string, roQuota v1.ResourceList) *peeringInfo {
+	peeringInfo, found := cache.getPeeringFromCache(clusterID)
+	if !found {
+		webhooklog.Info(fmt.Sprintf("\t\tPeeringInfo not found for clusterID %s", clusterID))
+		peeringInfo = createPeeringInfo(clusterID, clusterName, roQuota)
+		webhooklog.Info(fmt.Sprintf("\t\tPeeringInfo created for clusterID %s", clusterID))
+		webhooklog.Info(quotaFormatter(peeringInfo.getQuota(), "\t\tNew PeeringInfo Quota limits"))
+		webhooklog.Info(quotaFormatter(peeringInfo.getUsedQuota(), "\t\tNew PeeringInfo UsedQuota limits"))
+		webhooklog.Info(quotaFormatter(peeringInfo.getFreeQuota(), "\t\tNew PeeringInfo FreeQuota limits"))
+		cache.addPeeringToCache(clusterID, peeringInfo)
+	}
+	return peeringInfo
```

Comment:
This is not guaranteed to be atomic. You should use the `LoadOrStore` method provided by the SyncMap.